completion_aggregator/serializers.py

```diff
@@ -89,13 +96,30 @@ def __init__(self, user, course_key, queryset=None):
         """
         Initialize the adapter.
 
-        Optionally, an initial collection of aggregators may be provided.
+        Optionally, an initial collection of aggregators may be provided, though these may be recalculated if the course
+        is found to have stale completions.
         """
         self.user = user
         self.course_key = course_key
         self.aggregators = defaultdict(list)
-        if queryset:
-            self.update_aggregators(queryset)
+
+        # Check for stale completions, to trigger recalculating the aggregators if any are found.
+        with transaction.atomic():
+            stale_completions = [
+                stale.id for stale in StaleCompletion.objects.select_for_update().filter(
+                    resolved=False,
+                    username=self.user.username,
+                    course_key=self.course_key,
+                )
+            ]
+            is_stale = len(stale_completions) > 0
+
+            if is_stale:
+                log.info("Resolving %s stale completions for %s+%s",
+                         len(stale_completions), self.user.username, self.course_key)
+                StaleCompletion.objects.filter(id__in=stale_completions).update(resolved=True)
```

Comment:
Chose to resolve these `StaleCompletion` objects *prior* to running the `update_aggregators` step below, to prevent locking the database for too long here.